### PR TITLE
[SPARK-31894][SS] Introduce UnsafeRow format validation for streaming state store

### PR DESCRIPTION
**Pull request by xuanyuanking**
### What changes were proposed in this pull request?
Introduce UnsafeRow format validation for streaming state store.

### Why are the changes needed?
Currently, Structured Streaming directly puts the UnsafeRow into StateStore without any schema validation. It's a dangerous behavior when users reusing the checkpoint file during migration. Any changes or bug fix related to the aggregate function may cause random exceptions, even the wrong answer, e.g SPARK-28067.

### Does this PR introduce _any_ user-facing change?
Yes. If the underlying changes are detected when the checkpoint is reused during migration, the InvalidUnsafeRowException will be thrown.

### How was this patch tested?
UT added. Will also add integrated tests for more scenario in another PR separately.

### CODE REVIEW

**Comment by xuanyuanking** [comment]
cc @zsxwing @rednaxelafx @cloud-fan @HeartSaVioR

**Review by HeartSaVioR — COMMENTED** [review]
I agree this change is valid based on the fact Spark doesn't store the schema of state (and there's no validation between actual schema and the actual row), but this should be considered as a last resort because of the huge limitations. Safety guards must be placed in front of this - like SPARK-27237, which I think it covers various general issues with providing clearer guide of schema incompatibility between state and the query being run.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (HeartSaVioR) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1548]: BTW should we have configuration for this, given that this only does essential check which all rows must have been passed?

**Comment by skambha** [comment]
@xuanyuanking , can you please explain how this will fix the issue where we have changed something in the internal implementation of sum in SPARK-28067, how does that affect previous states and what would be the expected behavior.  From a query level, the sum schema is same.  Is the checkpoint storing information that is coming from intermediate states.   Are we storing unsafe rows from the updateExpression/ or the merge phases of aggregation?

**Comment by HeartSaVioR** [comment]
And personally I'd rather do the check in StateStore with additional overhead of reading "a" row in prior to achieve the same in all stateful operations. 

```
  /** Get or create a store associated with the id. */
  def get(
      storeProviderId: StateStoreProviderId,
      keySchema: StructType,
      valueSchema: StructType,
      indexOrdinal: Option[Int],
      version: Long,
      storeConf: StateStoreConf,
      hadoopConf: Configuration): StateStore = {
    require(version >= 0)
    val storeProvider = loadedProviders.synchronized {
      startMaintenanceIfNeeded()
      val provider = loadedProviders.getOrElseUpdate(
        storeProviderId,
        StateStoreProvider.createAndInit(
          storeProviderId.storeId, keySchema, valueSchema, indexOrdinal, storeConf, hadoopConf)
      )
      reportActiveStoreInstance(storeProviderId)
      provider
    }
    val store = storeProvider.getStore(version)
    val iter = store.iterator()
    if (iter.nonEmpty) {
      val rowPair = iter.next()
      val key = rowPair.key
      val value = rowPair.value
      // TODO: validate key with key schema
      // TODO: validate value with value schema
    }
    store
  }
```

For streaming aggregations it initializes "two" state stores so the overhead goes to "two" rows, but I don't think the overhead matters much.

**Comment by cloud-fan** [comment]
@skambha  it doesn't fix the issue, it gives a better error message when we hit the issue.

**Comment by cloud-fan** [comment]
I think this PR and SPARK-27237 are orthogonal, and we should have both. SPARK-27237 is a bit hard to be merged as it changes the checkpoint. We may need more reviews to see if it's future proof (e.g. when we want to support schema evolution of the state store).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingAggregationStateManager.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingAggregationStateManager.scala:119]: Actually that's the first version I did. Since the checking logic is only used for streaming aggregation query and also depends on the streaming config, I choose to put it in StreamingAggregationStateManager, WDYT?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingAggregationStateManager.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingAggregationStateManager.scala]: Thanks for the comments, I rephrase the error message to make it clearer. Yep, there are several ways that can lead to the invalid format and we need to list them all. Done in ee048bc

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Thanks, rename it in ee048bc. Considering it's an extra checking and still have overhead, I keep the feature flag for safety.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1550]: nit: `When true, check if the UnsafeRow from the state store is valid or not when running streaming aggregation queries. This can happen if the state store format has been changed.`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1553]: we usually put `internal()` right after `buildConf(...)`

**Comment by xuanyuanking** [comment]
> @skambha it doesn't fix the issue, it gives a better error message when we hit the issue.

Yep, I WIP for the integrated test of the state store format invalidation. I will show you the difference with/ this patch on the error message.

> Safety guards must be placed in front of this - like SPARK-27237, which I think it covers various general issues with providing clearer guide of schema incompatibility between state and the query being run.

Yes, the two approach addresses different sides of this issue, SPARK-27237 require an extra file to keep the schema, which can make the schema checking possible. This one is a guard for random failure or correctness bug.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Thanks, rephrase in 10a7980.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1550]: Thanks, done in 10a7980.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingAggregationStateManager.scala
Comment (rednaxelafx) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingAggregationStateManager.scala:119]: I was hoping we could move the core validation logic to either `UnsafeRow` itself, or some sort of `UnsafeRowUtils`, maybe somewhere in `sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util`.

This util function would either return a boolean indicating passed/failed integrity check, or it could return more details. I'd probably go with the former first. It would not do any conf checks -- that's the caller's responsibility. This utility is useful for debugging low-level stuff in general, and would come in handy in both Spark SQL and Structured Streaming debugging.

Then we can call that util function from here, after checking the confs. And the exception throwing logic can be left here too.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingAggregationStateManager.scala
Comment (rednaxelafx) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingAggregationStateManager.scala:129]: `UnsafeRow.calculateBitSetWidthInBytes(row.numFields) + 8 * row.numFields` this part is loop invariant. Please hoist it out of the loop manually here. It's the same kind of logic as `UnsafeRowWriter`'s
```java
    this.nullBitsSize = UnsafeRow.calculateBitSetWidthInBytes(numFields);
    this.fixedSize = nullBitsSize + 8 * numFields;
```
We may want to use the same or similar names for the hoisted variables.

`row.getSizeInBytes` on the next line is also loop invariant. Let's also hoist that out.

**Review by rednaxelafx — COMMENTED** [review]
Thank you very much for taking on this verification! I've used the same technique of checking unsafe row's structural integrity on quite a few occasions and it's been a very useful thing to have in the toolbox.

Having a util function that does this check inside of Spark would be very handy for future low-level debugging / investigations.

I agree with @cloud-fan that this verification feature is completely orthogonal to baking the schema info into the persisted state.

I'd strongly vote for having the schema info as a part of the persisted state instead of only having a blob that we interpret as `UnsafeRow`s without any guardrails. But doing so changes the binary format of the persisted state, so I'd really love to see it as a piece of the puzzle in a long term plan to improve the state store.

The verification proposed in this PR does not make any changes to the binary format, so it could be useful for both Spark master branch and existing releases.

---

For the `UnsafeRow` structural integrity guarantees / heuristics, I'd propose the following candidate invariants to consider: given a `row: UnsafeRow` and a `expectedSchema: StructType`
- `schema.fields.length == row.numFields` should always be true (already covered in this PR)
- `UnsafeRow.calculateBitSetWidthInBytes(row.numFields) <= row.getSizeInBytes` should always be true. A stricter `<` should be true if the `expectedSchema` contains at least one field. Not covered in this PR yet
- For variable-length fields: if null bit says it's null then don't do anything, else extract `offset` and `size`
  - `0 <= size < row.getSizeInBytes` should always be true. We can be even more precise than this, where the upper bound of `size` can only be as big as the variable length part of the row.
  - `offset` should be >= fixed sized part of the row
  - `offset + size` should be within the row bounds (already covered by this PR)
  - We can make further assumptions on the `UnsafeRow` format, by assuming that if `field1.ordinal < field2.ordinal`, then `field1.offset + field1.size <= field2.offset`. This assumes that the fields were written into in left-to-right order, which doesn't have to be the case, but all the write logic I know of in Spark fits this assumption. So this can be considered an optional heuristic.
- For fixed-length fields that are narrower than 8 bytes (boolean / byte / short / int / float), if null bit says it's null then don't do anything, else:
  - check if the unused bits in the field are all zeros. The `UnsafeRowWriter`'s `write()` methods make this guarantee.

When I did manual debugging, sometimes I'd also check the first couple of characters in a `UTF8String` from an `UnsafeRow` and see if the characters make sense as UTF-8. That's not something easily checkable here so I wouldn't suggest that.

If we know the length of the entire buffer of the backing store for `UnsafeRow`s, we should make sure our `offset + size` never goes beyond that, too.

**Comment by cloud-fan** [comment]
> Having a util function that does this check inside of Spark would be very handy for future low-level debugging / investigations.

+1. How about put the validation code in a new object `UnsafeRowUtils`?

**Comment by xuanyuanking** [comment]
@rednaxelafx Great thanks for the detailed comment and guidance. I'm addressing these comments.

**Comment by HeartSaVioR** [comment]
Will this be included to Spark 3.0.0? If this is to unblock SPARK-28067 to be included to Spark 3.0.0 then it's OK to consider this first, but if this plans to go to Spark 3.1 then I'm not sure about the priority - are all of you aware that the PR for SPARK-27237 was submitted more than a year ago, and still be considered as later?

I still don't get why the proposal is restricting its usage to streaming aggregation, whereas the mechanism is validation of the UnsafeRow which can be applied to all stateful operations. Let's not to pinpoint the problem we've just seen.

Also from my side the overhead of the validation logic looks to be trivial compared to the operations stateful operators will take - we don't do the validation for all rows, even don't sample, just the first one. Unless we have a chance to bring a show-stopper bug in the validation logic (so that we need to provide the way to disable the validation), I'm not seeing the needs of new configuration.

**Comment by HeartSaVioR** [comment]
And I think SPARK-27237 doesn't require a sort of "future-proof" which is preferably be done with a thing with risk - it doesn't touch the existing part of checkpoint and simply put the schema information into a new file. If we find a better way to pack the schema information into the checkpoint, we can simply discard/ignore the file or craft a logic to migrate smoothly. No risk on rolling back in future.

**Comment by cloud-fan** [comment]
Yea we need this PR to unblock backporting SPARK-28067 to 3.0.

> the mechanism is a validation of the UnsafeRow which can be applied to all stateful operations.

What are other stateful operations that use unsafe row? I think we can apply the check everywhere.

> Unless we have a chance to bring a show-stopper bug in the validation logic

This is something we don't know. Adding a flag seems safer.

> it doesn't touch the existing part of checkpoint and simply put the schema information into a new file.

I'm not saying we shouldn't merge it. I just want to prioritize this PR so that we may be able to include sum correctness bug in 3.0.

**Comment by HeartSaVioR** [comment]
> What are other stateful operations that use unsafe row? I think we can apply the check everywhere.

State store itself stores UnsafeRow, hence it applies to everywhere in stateful operations. I'd propose to do it like https://github.com/apache/spark/pull/28707#issuecomment-637926400 instead of fixing everywhere.

**Comment by xuanyuanking** [comment]
@skambha You can check the integrated tests in #28725. If we delete the validation, we'll get a NPE for [this test](https://github.com/apache/spark/pull/28725/files#diff-492f0d70824a58ef2ea94a54dc6f9707R79), and get an assertion in the unsafe row for [this test](https://github.com/apache/spark/pull/28725/files#diff-492f0d70824a58ef2ea94a54dc6f9707R185). That is to say, we will get random failures during reusing the checkpoint written by the old Spark version.

**Comment by xuanyuanking** [comment]
@HeartSaVioR After taking a further look. Instead of dealing with the iterator, how about adding the invalidation for all state store operations in `StateStoreProvider`? Since we can get the key/value row during load map. WDYT?

**Comment by skambha** [comment]
> @skambha You can check the integrated tests in #28725. If we delete the validation, we'll get a NPE for [this test](https://github.com/apache/spark/pull/28725/files#diff-492f0d70824a58ef2ea94a54dc6f9707R79), and get an assertion in the unsafe row for [this test](https://github.com/apache/spark/pull/28725/files#diff-492f0d70824a58ef2ea94a54dc6f9707R185). That is to say, we will get random failures during reusing the checkpoint written by the old Spark version.

Thanks for adding the test.

**Comment by HeartSaVioR** [comment]
> @HeartSaVioR After taking a further look. Instead of dealing with the iterator, how about adding the invalidation for all state store operations in StateStoreProvider? Since we can get the key/value row during load map. WDYT?

It would be nice to see the proposed change by code to avoid misunderstanding, like I proposed in previous comment. (anything including commit in your fork or text comment is OK) I'll try out my alternative (wrapping State Store) and show the code change. Thanks!

**Comment by HeartSaVioR** [comment]
My alternative with wrapping state store is something like below:

```
  class RowValidatingStateStore(
      underlying: StateStore,
      keyType: Seq[DataType],
      valueType: Seq[DataType]) extends StateStore {
    private var isValidated = false

    override def get(key: UnsafeRow): UnsafeRow = {
      val value = underlying.get(key)
      if (!isValidated) {
        validateRow(value)
        isValidated = true
      }
      value
    }

    override def id: StateStoreId = underlying.id
    override def version: Long = underlying.version
    override def put(key: UnsafeRow, value: UnsafeRow): Unit = underlying.put(key, value)
    override def remove(key: UnsafeRow): Unit = underlying.remove(key)
    override def commit(): Long = underlying.commit()
    override def abort(): Unit = underlying.abort()
    override def iterator(): Iterator[UnsafeRowPair] = underlying.iterator()
    override def metrics: StateStoreMetrics = underlying.metrics
    override def hasCommitted: Boolean = underlying.hasCommitted

    private def validateRow(row: UnsafeRow): Unit = {
      // TODO: call util method with row and schema to validate
    }
  }

  def get(...): StateStore = {
    require(version >= 0)
    val storeProvider = loadedProviders.synchronized {
      ...
    }
    // TODO: add if statement to see whether it should wrap state store or not
    new RowValidatingStateStore(storeProvider.getStore(version, keySchema, valueSchema))
  }
```

The example code only checks in get operation, which is insufficient to check "key" row in state. That said, iterator approach still provides more possibility of validation, though the validation of unsafe row itself doesn't have enough coverage of checking various incompatibility issues (Definitely we should have another guards as well) so that's a sort of OK to only cover value side.

**Comment by xuanyuanking** [comment]
All the comments addressed in 1f71563. Thanks for the review!
It also includes my alternative of adding the invalidation for all state store operations in StateStoreProvider, PTAL.

**Comment by HeartSaVioR** [comment]
Sorry my comment was edited so you may be missed the content, but it is also a sort of pointing out of "pinpoint" - do you think your approach works with other state store providers as well? The root cause isn't bound to the implementation of state store provider but this patch is only addressing in HDFS state store provider.

I guess you're trying to find how it can be done less frequently, first time the state is loaded from the file, which is optimal. While I think it can be even done without binding to the state store provider implementation if we really need it, have we measured the actual overhead? If the overhead turns out to be trivial then it won't be matter we run validation check for each batch. It sounds to be sub-optimal, but the overhead would be trivial.

**Comment by xuanyuanking** [comment]
A new regression bug SPARK-31990 was found when investigating the test failure https://github.com/apache/spark/pull/28707#issuecomment-639861273. The root cause is that [this line](https://github.com/apache/spark/pull/28062/files#diff-7a46f10c3cedbf013cf255564d9483cdL2458) in SPARK-31292 made the order of groupCols in Deduplicate changed, and the order changing will break the validation logic here. That is to say, if we don't have this PR, the executor JVM could probably crash, throw a random exception, or even return a wrong answer when using the checkpoint written by the previous version.

So we have 2 related work of this PR:

- [ ] Fix and merge the compatibility issue in #28830
- [ ] Add new test(or modify the current Kafka test) in #28725

------------------
### More detailed analysis:
The expected order of `Deduplicate.groupCols` in UT KafkaMicroBatchV2SourceSuite is
```
[timestamp, partition, timestampType, key, offset, topic, value]
```
After the changes in SPARK-31292, the groupCols changed to
```
[key, value, topic, partition, offset, timestamp, timestampType]
```

#### Why this incompatibility bug didn't fail the `KafkaMicroBatchV2SourceSuite` when it merged?

Because the UT `default config of includeHeader doesn't break the existing query from Spark 2.4` didn't test the scenario of duplicating and check the answer.
Although the UT uses the checkpoint written by version 2.4.3 and streaming duplicate operation, it just wants to prove that the new header(added in SPARK-23539) doesn't break the original checkpoint file.

**Comment by xuanyuanking** [comment]
```
Sorry my comment was edited so you may be missed the content, but it is also a sort of pointing out for "pinpointing" - do you think your approach works with other state store providers as well?
```
@HeartSaVioR I totally agree with your approach is clean enough for all the state store provider, that's also what I want to achieve. But as you may find that the new regression bug SPARK-31990 is found by key validation, that's why we can't directly skip it. Anyway, super thanks for the proposal, let's try to find another way together which matches all the requirements, without changing the provider API.

**Comment by cloud-fan** [comment]
I guess we may need to change the state store API to allow Spark to do validation easily. I'm OK to put the validation logic in the state store implementation for now, and figure out how to change the state store API later.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/UnsafeRowUtils.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/UnsafeRowUtils.scala:77]: For null field, we can also check `row.getLong(index) == 0`, as it's also guaranteed by the `UnsafeRowWriter`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala:155]: `For the first case`: I think it's for the cases?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala:152]: Does it have  to be `SparkException`?

**Review by cloud-fan — APPROVED** [review]
LGTM except for a few comments.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala:262]: Can we add a TODO that this validation should be moved to a higher level so that it works from all state store implementations?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/UnsafeRowUtils.scala
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/UnsafeRowUtils.scala:79]: Thanks, done in fd74ff9.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala:154]: The resolution is for the first case. For the rest cases listing, they should be considered as user problems.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala]: No, change it to RuntimeException. Done in fd74ff9.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala
Comment (xuanyuanking) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala:264]: Thanks, add the TODO in fd74ff9.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/UnsafeRowUtils.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/UnsafeRowUtils.scala:77]: I mean `case (field, index) if row.isNullAt(index)`

`UnsafeRowWriter` will zero-out the offsetAndRegion if the field value is null.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (HeartSaVioR) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1244]: I'd explicitly describe that the feature is only effective in built-in HDFS state store provider.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/UnsafeRowUtils.scala
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/UnsafeRowUtils.scala]: Thanks for the explanation, done in 557eb30.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Make senes, done in 557eb30.

**Comment by cloud-fan** [comment]
thanks, merging to master! (I think this patch is too big to backport)

**Comment by xuanyuanking** [comment]
Thanks all for reviewing!
I'll review #24173 as the next step for schema validation.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1243]: Change `UnsafeRow` to `checkpoint` ? Most end users do not know what are `UnsafeRow`

Comment (xuanyuanking) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1243]: Sure, will submit a follow-up PR today.

**Pull request closed without merging**
